### PR TITLE
Noise: `PatternXX` and testing with go

### PR DESCRIPTION
**Pull request by mhchia**
## What was wrong?

#375 

## How was it fixed?
Based on #405 

* Feature
    * Noise
        * Handshake
            * `Pattern XX`
        * Encryption/decryption
            * `SecureSession`(a buffered stream reader originally used in secio) is refactored and used by both `Noise` and `Secio` now.
* Refactor
    * msgio: dedup the original similar implementations.
        * `MsgReadWriter`
        * `FixedSizeLenMsgReadWriter`
            * `NoiseReadWriter`
        * `VarIntLengthMsgReadWriter`
            * `PlaintextHandshakeReadWriter`
* Fix
    * Plaintext: use `VarIntLengthMsgReadWriter` over `FixedSizeLenMsgReadWriter`. Ref: https://github.com/libp2p/go-libp2p-core/pull/74/
* Tests
    * Transport, encryption/decryption, security multistream.
    * Interop test with go implementation
        * through the modified [`go-libp2p-daemon`](https://github.com/mhchia/go-libp2p-daemon/commit/558b58db41c1821ce9d2537235ca1bbe0030ce3e) against different security protocol.
            * Plaintext
            * Noise: [`go-libp2p-noise`](https://github.com/libp2p/go-libp2p-noise)
                * Issue: msg#3(initiator sends the payload to the responder) is different from the one in go implementation. Workaround this through ignoring the first `32` bytes for now. Will investigate what is wrong later.
* Misc
    * Re-compile protobufs
        * `noise.proto`
        * `plaintext.proto`

### TODO
- [ ] Confirm why msg#3 is different between py and go.
- [ ] Add more exceptions and handling.
- [ ] Add more comments.

### Reference
- [noise spec](https://noiseprotocol.org/noise.html#message-format)
- [noise-libp2p spec](https://github.com/libp2p/specs/tree/master/noise#xx)
- [go-libp2p-noise](https://github.com/libp2p/go-libp2p-noise)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by mhchia** [comment]
Though there are things to be added and confirmed, I think this should be ready for review. We can talk with go implementation through `noise` now. 🎉

File: libp2p/io/msgio.py [hunk 1]
```diff
@@ -5,80 +5,85 @@
...
-async def read_length(reader: Reader) -> int:
-    length_bytes = await read_exactly(reader, SIZE_LEN_BYTES)
+async def read_length(reader: Reader, size_len_bytes: int) -> int:
+    length_bytes = await read_exactly(reader, size_len_bytes)
     return int.from_bytes(length_bytes, byteorder=BYTE_ORDER)
 
 
-def encode_msg_with_length(msg_bytes: bytes) -> bytes:
-    len_prefix = len(msg_bytes).to_bytes(SIZE_LEN_BYTES, "big")
+def encode_msg_with_length(msg_bytes: bytes, size_len_bytes: int) -> bytes:
```
Comment (pipermerriam) [inline libp2p/io/msgio.py]: Why is the size passed in here instead of measuring it from `len(msg_bytes)`?

File: libp2p/io/msgio.py [hunk 2]
```diff
@@ -5,80 +5,85 @@
...
 
+    @abstractmethod
+    async def next_msg_len(self) -> int:
+        ...
 
-class MsgIOReader(ReadCloser):
-    read_closer: ReadCloser
-    next_length: Optional[int]
+    @abstractmethod
+    def encode_msg(self, msg: bytes) -> bytes:
```
Comment (pipermerriam) [inline libp2p/io/msgio.py]: should this be public?  Is it used as an external API or only called internally?

Comment (mhchia) [reply libp2p/io/msgio.py]: It's only called internally, but I expect subclasses to implement it to make `write_msg ` work. So as `next_msg_len` to make `read_msg` work.

File: libp2p/io/msgio.py (hunk 1 above)
Comment (mhchia) [reply libp2p/io/msgio.py]: `size_len_bytes` means the size of the length field in the message. `len(msg_bytes)` is encoded and put in the length field.

**Comment by mhchia** [comment]
I'm merging this PR since it gets too large and it already works fine with the go implementation. Will open a new one if everything gets updated.

**Pull request merged**